#! /bin/bash

# Training AUGUSTUS requires scripts and programs from MAKER, BLAT, UCSC Utils, AUGUSTUS.

# MAKER - http://www.yandell-lab.org/software/maker.html
# AUGUSTUS - http://bioinf.uni-greifswald.de/augustus/
# BLAT - http://genome.ucsc.edu/FAQ/FAQblat.html
# UCSC Utils - http://hgdownload.cse.ucsc.edu/admin/exe/

# Additionally, the script, fathom_to_genbank.pl, is required and can be found at github: https://github.com/Childs-Lab/GC_specific_MAKER.
# This script must be available in your $PATH.
# This script requires BioPerl.

# Kevin Childs

WORKING_DIR=$1
DATASTORE_INDEX_LOG=$2
AUGUSTUS_SPECIES_NAME=$3
GENOME_FASTA=$4
CDNA_FASTA=$5


if [[ -z $WORKING_DIR || -z $DATASTORE_INDEX_LOG || -z $AUGUSTUS_SPECIES_NAME || -z $GENOME_FASTA || -z $CDNA_FASTA ]]; then
    echo "Usage: $0 WORKING_DIR  DATASTORE_INDEX_LOG  AUGUSTUS_SPECIES_NAME  GENOME_FASTA  CDNA_FASTA"
    exit 1
elif [ ! -e $DATASTORE_INDEX_LOG ]; then
    echo "The previous index log file does not exist: $DATASTORE_INDEX_LOG"
    exit 1
elif [ ! -e $GENOME_FASTA ]; then
    echo "The genome fasta file does not exist: $GENOME_FASTA"
    exit 1
elif [ ! -e $CDNA_FASTA ]; then
    echo "The cDNA fasta file does not exist: $CDNA_FASTA"
    exit 1
fi


# 1. Convert the maker output to zff format using the maker datastore log file

echo 'cd $WORKING_DIR'
cd $WORKING_DIR

echo 'maker2zff -x 0.2 -l 200 -d $DATASTORE_INDEX_LOG'
maker2zff -x 0.2 -l 200 -d $DATASTORE_INDEX_LOG


# output: genome.ann (ZFF file)
#         genome.dna (fasta file that the coordinates in the zff can be referenced againsted)

# 2. use the fathom  script from snap to get the unique annotations

echo 'fathom genome.ann genome.dna -categorize 1000'
fathom genome.ann genome.dna -categorize 1000

NUMFOUND="`grep -c '>' uni.ann`"

if [ ${NUMFOUND} -gt 499 ]; then
    NUMFOUND=500
fi

TEMPSPLIT=$((NUMFOUND/2))
NUMSPLIT=${TEMPSPLIT/.*}

echo "number found after fathom: $NUMFOUND"
echo "number after split: $NUMSPLIT"

# 3. Convert the uni.ann and uni.dna output from fathom into a genbank formatted file.
#    fathom_to_genbank.pl is available on github: https://github.com/Childs-Lab/GC_specific_MAKER.
#    Modify the path here, or ensure that fathom_to_genbank.pl is in your $PATH.

echo 'fathom_to_genbank.pl  --annotation_file uni.ann  --dna_file uni.dna  --genbank_file augustus.gb  --number ${NUMFOUND}'
fathom_to_genbank.pl  --annotation_file uni.ann  --dna_file uni.dna  --genbank_file augustus.gb  --number ${NUMFOUND}


# 4. Split the known genes into test and training files.
#    randomSplit.pl is a script provided by AUGUSTUS.
#    Modify the path here, or ensure that randomSplit.pl is in your $PATH.

echo 'randomSplit.pl ${WORKING_DIR}/augustus.gb ${NUMSPLIT}'
randomSplit.pl ${WORKING_DIR}/augustus.gb ${NUMSPLIT}

# We will use autoAug.pl for training because we have transcript alignments that will be used as hints.
# The etraining and optimize_augustus.pl scripts from AUGUSTUS will not be used as they do not allow for the use of hints.
# 5. Run autoAug.pl.  
#    This script is provided in the AUGUSTUS installation.
#    Modify the path here, or ensure that fathom_to_genbank.pl is in your $PATH.

echo 'autoAug.pl --species=$AUGUSTUS_SPECIES_NAME --genome=$GENOME_FASTA --trainingset=${WORKING_DIR}/augustus.gb.train --cdna=$CDNA_FASTA  --noutr'
autoAug.pl --species=$AUGUSTUS_SPECIES_NAME --genome=$GENOME_FASTA --trainingset=${WORKING_DIR}/augustus.gb.train --cdna=$CDNA_FASTA  --noutr

# 6. Run the batch scripts generated by autoAug.pl

echo 'cd ${WORKING_DIR}/autoAug/autoAugPred_abinitio/shells'
cd ${WORKING_DIR}/autoAug/autoAugPred_abinitio/shells

# The number of ./aug# scripts is variable.  Run until the new file does not exist.
x=1
while [ -e ./aug${x} ]
do
    ./aug${x}
    let x=x+1
done


# 7. Checked sensitivity and specificity of the newly trained AUGUSTUS HMM by using the test data.

echo 'augustus --species=$AUGUSTUS_SPECIES_NAME augustus.gb.test'
augustus --species=$AUGUSTUS_SPECIES_NAME augustus.gb.test

